File: leetcode/problems/medium/3.longest_substring_without_repeating_characters/python3/solution.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
# Runtime 89ms Beats 31.78% Memory 16.3MB Beats 61.75%
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 2:
            return len(s)
        start, end = 0, 0
        substring = set()
        max_count = 0
        while end < len(s):
            if s[end] not in substring:
                substring.add(s[end])
                end += 1
                max_count = max(max_count, len(substring))
            else:
                substring.remove(s[start])
                start += 1
            
        return max_count

File: leetcode/problems/medium/3.longest_substring_without_repeating_characters/kotlin/Solution.kt
// https://leetcode.com/problems/longest-substring-without-repeating-characters/?envType=problem-list-v2&envId=rab78cw1
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val letters = BooleanArray(128)
        var maxLength = 0
        var start = 0
        for (i in s.indices) {
            val letterIndex = s[i].code
            if (letters[letterIndex]) {
                while (s[i] != s[start]) {
                    letters[s[start].code] = false
                    start++
                }
                start++
            }
            letters[letterIndex] = true
            maxLength = maxOf(maxLength, i - start + 1)
        }

        return maxLength
    }
}
